demos/defender/defender_demo_json/metrics_collector.c

```diff
@@ -378,4 +378,152 @@ MetricsCollectorStatus_t GetEstablishedConnections( Connection_t * pOutConnectio
+    FILE * fileHandle = NULL;
+    /* Variables for reading and processing data from "/proc/meminfo" file. */
+    uint8_t lineNumber = 0;
+    char lineBuffer[ MAX_LINE_LENGTH ];
+    bool readTotalMem = false, readAvailableMem = false;
+    const char * const pTotalMemoryField = "MemTotal";
+    const char * const pAvailableMemField = "MemAvailable";
+    int filledVariables = 0;
+    uint64_t parsedMemData = 0UL;
+
+    if( ( pMemoryData == NULL ) )
+    {
+        LogError( ( "Invalid parameter. pMemoryData=%p", ( void * ) pMemoryData ) );
+        status = MetricsCollectorBadParameter;
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        fileHandle = fopen( "/proc/meminfo", "r" );
+
+        if( fileHandle == NULL )
+        {
+            LogError( ( "Failed to open /proc/meminfo." ) );
+            status = MetricsCollectorFileOpenFailed;
+        }
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        while( ( fgets( &( lineBuffer[ 0 ] ), MAX_LINE_LENGTH, fileHandle ) != NULL ) )
+        {
+            lineNumber++;
+
+            LogDebug( ( "File: /proc/meminfo, LineNo: %u, Content: %s.",
+                        lineNumber, &( lineBuffer[ 0 ] ) ) );
+
+            /* Extract the total memory value from the line. */
+            filledVariables = sscanf( lineBuffer,
+                                      "%*[^1-9]%lu kB",
+                                      ( &parsedMemData ) );
+
+            if( filledVariables != 1 )
+            {
+                LogError( ( "Failed to parse data. File: /proc/meminfo, Line:%d, Content:%s", lineNumber, lineBuffer ) );
+                status = MetricsCollectorParsingFailed;
+            }
+            /* Check if the line read represents information for "total memory" in the system. */
+            else if( strncmp( lineBuffer, pTotalMemoryField, strlen( pTotalMemoryField ) ) == 0 )
+            {
+                /* Populate buffer with information for total memory as "<data>kB". */
+                ( void ) snprintf( pMemoryData->totalMemory, sizeof( pMemoryData->totalMemory ),
+                                   "%lukB", parsedMemData );
+
+                readTotalMem = true;
+            }
+            /* Check if the line read represents information for "available memory" in the system. */
+            else if( strncmp( lineBuffer, pAvailableMemField, strlen( pAvailableMemField ) ) == 0 )
+            {
+                /* Populate buffer with information for total memory as "<data>kB". */
+                ( void ) snprintf( pMemoryData->availableMemory, sizeof( pMemoryData->availableMemory ),
+                                   "%lukB", parsedMemData );
+
+                readAvailableMem = true;
+            }
+
+            if( readTotalMem && readAvailableMem )
+            {
+                /* We have obtained data for both total and available memory in the system. */
+                break;
+            }
+        }
+    }
+
```

Comment:
If no line starts with `MemTotal` or `MemAvailable`, the function will return success without ever filling the output buffers?

Reply:
Good catch, have updated to return an error code in that case.

Reply:
It may be better to move this assignment out of loop:

```c
if( readTotalMem && readAvailableMem )
{
    /* We have obtained data for both total and available memory in the system. */
    status = MetricsCollectorSuccess;
}
else
{
    status = MetricsCollectorDataNotFound;
}
```